src/checkers/inference/model/TernaryConstraint.java

```diff
@@ -0,0 +1,31 @@
+package checkers.inference.model;
+
+/**
+ * This interface is implemented by constraints between three variables. Make is provided to be able
+ * to make a copy of the constraint without knowing the concrete base class. {@link #getFirst()},
+ * {@link #getSecond()}, and {@link #getThird()} are provided so that users can either copy through
+ * a slot to a constraint made by make or substitute out a variable. E.g.,
+ *
+ * <p>
+ * This call makes a copy of the constraint:
+ * {@code make(getFirst(), getSecond(), getThird(), getLocation()}
+ *
+ * <p>
+ * This call modifies the first slot:
+ * {@code make(mutate(getFirst()), getSecond(), getThird(), getLocation())}
+ * 
+ * <p>
+ * Note: {@code getLocation()} is provided by {@link Constraint}
+ */
+public interface TernaryConstraint {
```

Comment:
Follow the javadoc of this class, it seems the main intention of this interface is:
"to be able to make a copy of the constraint without knowing the concrete base class"

If so, why restrict this on `TernaryConstraint`?

I can see the copy is order-sensitive, but since we already have `getSlots()` method defined in `Constraint` class, so maybe we can do it more general:

```java
public interface MakableConstraint {
Constraint make(List<Slot> slots, final AnnotationLocation location);
}
```

Reply:
The definition is a clone of [BinaryConstraint](https://github.com/opprop/checker-framework-inference/blob/master/src/checkers/inference/model/BinaryConstraint.java) with 3 slots instead of 2, which has the make method.

As far as I know, the make method is called by the ExistentialVariableNormalizer but i'm not intricately familiar with what that does.

The only goal here is to provide commonality in the definitions, in case ternary constraints some how get normalized in the future.

Reply:
The make() allows reassignment of the final fields in BinaryConstraint, whereas clone() doesn't allow it. Overall the removal of make() (and rewriting of how ExistentialVariables are processed) is a separate PR.

Reply:
> The make() allows reassignment of the final fields in BinaryConstraint

Doesn't sound problematic for me: if a field is final, how can you reassign that?

I didn't see there is field reassignment either in body of `make()` or after calling `make()`.

Reply:
According to the use in ConstraintNormalizer, the point of make() is to create the same kind of constraint as the given `constraint` object but with different slots. The body of make in each of its implementations clearly create a brand new constraint of the same type, with the final fields initialized to the new slots. However, the call site in existentialTree doesn't have to do a series of instanceof checks to know which constraint type to "clone". I don't see how `clone()` can support this specific behavior that is in use in ConstraintNormalizer. But if you feel like removing make() all together from BinaryConstraint, feel free to do that in a separate PR since that requires reimplementing ConstraintNormalizer.

I'll get rid of make() in TernaryConstraint. Overall, what your feedback suggests is that prior design decisions (in BinaryConstraint) aren't thoroughly examined.